083-stretch-cluster.md

```diff
@@ -0,0 +1,155 @@
+
+### Design
+
+The cluster operator will be deployed in all Kubernetes clusters and will manage Kafka brokers/controllers running on that cluster. One Kubernetes cluster will act as the control point for defining custom resources (Kafka, KafkaNodePool) required for stretch Kafka cluster. The KafkaNodePool custom resource will be extended to include information about a Kubernetes cluster where the pool should be deployed. The cluster operator will create necessary resources (StrimziPodSets, services etc.) on the target clusters specified within the KafkaNodePool resource.
+
+This approach will allow users to specify/manage the definition of stretch Kafka cluster in a single location. The operators will then create necessary resources in target Kubernetes clusters, which can then be reconciled/managed by operators on those clusters.
+
+### Reconciling Kafka and KafkaNodePool resources
+<img width="1321" alt="image" src="https://media.github.ibm.com/user/473362/files/0a1a4e7f-8227-47da-86b1-4b27a83fdb9d">
+
+### Reconciling StrimziPodSet resources
+<img width="1317" alt="image" src="https://media.github.ibm.com/user/473362/files/f4524847-3c9c-4a8c-a5fd-4a76b6ec0bf4">
+
+#### KafkaNodePool changes
+A new optional field (`target`) will be introduced in the KafkaNodePool resource specification, to allow users to specify the details of the Kubernetes cluster where the node pool should be deployed. This section will include the target cluster's URL (Kubernetes cluster where resources for this node pool will be created) and the secret containing the kubeconfig data for that cluster. 
+
+An example of the KafkaNodePool resource with the new fields might look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  target:
+    clusterUrl: <K8S Cluster URL>
+    secret: <SecretName>
+  listenerConfig:
+    - configuration:
+          bootstrap:
+            alternativeNames:
+              - bootstrap-ext.<kubernetes infra host>
+              - broker-100.<kubernetes infra host>
+              - broker-101.<kubernetes infra host>
+            host: bootstrap-ext.<kubernetes infra host>
+          brokers:
+          - broker: 100
+            host: broker-100.<kubernetes infra host>
+          - broker: 101
+            host: broker-101.<kubernetes infra host>
+          class: nginx
+        name: connectext
+  roles:
+    - controller
+  storage:
+   .........
+```
+
+#### Kafka changes
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/node-pools: enabled
+    strimzi.io/kraft: enabled
+    strimzi.io/stretch-mode: enabled
```

Comment:
Thank you for the suggestion! May I ask why moving the `stretch-mode` annotation to the NodePools would be a good idea?

We think that adding the stretch-mode annotation in the Kafka CR makes sense because it clearly represents a global configuration that applies to the entire Kafka cluster. By placing it in the Kafka CR, it signals that the entire cluster is operating in stretch mode, affecting how brokers, controllers, and listeners are handled across multiple Kubernetes clusters.

Having this configuration at the Kafka level also makes it easier to manage and audit, as it is immediately visible from the main Kafka resource. This avoids scattering critical configurations across multiple NodePool resources, which could lead to complexity when maintaining or troubleshooting the cluster. Additionally, stretch mode is fundamentally a clusterwide behavior rather than something that is specific to individual node pools, so we believe the Kafka CR is the most appropriate place to define it.

Reply:
A pool cannot be stretched AFAIU from the proposal, so I think the annotation belongs to the `Kafka` custom resource. Having it on the node pool let me think that pods for the specific pool are stretched which should not be the case.

Reply:
I completely agree with your point. By keeping the annotation in the Kafka CR, we ensure that the stretch configuration remains cluster-wide, clearly indicating that it applies across all nodes and resources. This also simplifies the management and understanding of the cluster’s operational mode